File: src/main/java/solution/java/medium/n322/Solution.java
package solution.java.medium.n322;

/**
 * 322. Coin Change
 * <p>
 * You are given coins of different denominations and a total amount of money amount. Write a function to compute the
 * fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any
 * combination of the coins, return -1.
 */
public class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0) {
            return 0;
        }
        int[] dp = new int[amount];
        for (int i = 1; i <= amount; i++) {
            rec(coins, i, dp);
        }
        return dp[amount - 1];
    }

    public int rec(int[] coins, int amount, int[] dp) {
        if (amount < 0) {
            return -1;
        }
        if (amount == 0) {
            return 0;
        }
        if (dp[amount - 1] > 0) {
            return dp[amount - 1];
        }
        int result = Integer.MAX_VALUE;
        for (int coin : coins) {
            int rec = rec(coins, amount - coin, dp) + 1;
            if (rec > 0 && rec < result) {
                result = rec;
            }
        }
        dp[amount - 1] = result == Integer.MAX_VALUE ? -1 : result;
        return dp[amount - 1];
    }
}

File: src/main/kotlin/solution/medium/n322/Solution.kt
package solution.medium.n322

class Solution {
    fun coinChange(coins: IntArray, amount: Int): Int {
        if (amount < 1) {
            return 0
        }
        return getCount(coins, amount, IntArray(amount))
    }

    private fun getCount(coins: IntArray, amount: Int, dp: IntArray): Int {
        if (amount == 0) {
            return 0
        }
        if (amount < 0) {
            return -1
        }
        if (dp[amount - 1] != 0) return dp[amount - 1]
        var min = Int.MAX_VALUE
        for (coin in coins) {
            val next = getCount(coins, amount - coin, dp)
            if (next in 0 until min) {
                min = next + 1
            }
        }
        dp[amount - 1] = if (min == Int.MAX_VALUE) -1 else min
        return dp[amount - 1]
    }


}